Typescritp error after upgrading to 4.3.0

The error I am getting is: https://github.com/carloscortonc/cli-er/actions/runs/17402416557/job/49398408512?pr=98#step:6:7

```sh
Error: node_modules/tstyche/build/index.d.ts(103,48): error TS2304: Cannot find name 'DecoratorContext'.
```

After some digging, I found out this was introduced in v4.0.0 [here](https://github.com/tstyche/tstyche/pull/447/files)

https://github.com/tstyche/tstyche/blob/2039388bba7f19b3157b06735314604f3d06c83b/source/types.ts#L105

From what I could read, this type was introduced in typescript 5.0, but tstyche documentation states that only version `>=4.7` is required (https://tstyche.org/guides/typescript-versions#supported-versions)

--- a/package.json
+++ b/package.json
@@ -22,6 +22,10 @@
   "type": "module",
   "exports": {
     ".": {
+      "types@4.x": {
+        "import": "./build/4.x/index.d.ts",
+        "require": "./build/4.x/index.d.cts"
+      },
       "import": "./build/index.js",
       "require": "./build/index.cjs"
     },
@@ -30,6 +34,13 @@
   },
   "main": "./build/index.js",
   "types": "./build/index.d.ts",
+  "typesVersions": {
+    "4.x": {
+      "build/*": [
+        "./build/4.x/*"
+      ]
+    }
+  },
   "bin": "./build/bin.js",
   "files": [
     "build/*"

--- a/scripts/rollup.ts
+++ b/scripts/rollup.ts
@@ -77,6 +77,27 @@ function tidyDts(): Plugin {
   };
 }
 
+// TODO remove this plugin and 'package.json' entries after dropping support for TypeScript 4.x
+function ts4Dts(): Plugin {
+  return {
+    name: "ts4-dts",
+
+    async renderChunk(code, chunkInfo) {
+      if (chunkInfo.fileName.startsWith("index")) {
+        const updatedCode = code.replace(
+          / {4}\/\*\*\n {5}\* Checks if the decorator[\s\S]*DecoratorContext\) => void;\n/,
+          "",
+        );
+
+        await fs.mkdir(path.resolve(output.dir, "4.x"), { recursive: true });
+        await fs.writeFile(path.resolve(output.dir, "4.x", chunkInfo.fileName), updatedCode);
+      }
+
+      return null;
+    },
+  };
+}
+
 const config: Array<RollupOptions> = [
   {
     external: [/^node:/],
@@ -91,6 +112,7 @@ const config: Array<RollupOptions> = [
       typescript({ tsconfig }),
       dts({ tsconfig }),
       tidyDts(),
+      ts4Dts(),
     ],
   },
 
@@ -104,6 +126,7 @@ const config: Array<RollupOptions> = [
       // @ts-expect-error TODO: https://github.com/rollup/plugins/issues/1541
       typescript({ tsconfig }),
       dts({ tsconfig }),
+      ts4Dts(),
     ],
   },
 

